src/Client.php

```diff
@@ -117,152 +129,180 @@ public function getVersion(bool $without_hash = true): string
+            $this->doRequest(new Request('get', 'user'), ['query' => $query])
         );
     }
 
     /**
      * {@inheritdoc}
      */
-    public function userBalance(string $report_type_uid, bool $detailed = false): UserBalanceResponse
+    public function userBalance(UserBalanceParams $params): UserBalanceResponse
     {
+        $request_url = \sprintf('user/balance/%s', \urlencode($params->getReportTypeUid()));
+
+        $query = [
+            '_detailed' => $params->isDetailed() ? 'true' : 'false',
+        ];
+
         return UserBalanceResponse::fromHttpResponse(
-            $this->doRequest(new Request('get', \sprintf('user/balance/%s', \urlencode($report_type_uid))), [
-                'query' => [
-                    '_detailed' => $detailed
-                        ? 'true'
-                        : 'false',
-                ],
-            ])
+            $this->doRequest(new Request('get', $request_url), ['query' => $query])
         );
     }
 
     /**
      * {@inheritdoc}
      */
-    public function userReportTypes(bool $can_generate = false,
-                                    bool $content = false,
-                                    string $query = '_all',
-                                    int $size = 20,
-                                    int $offset = 0,
-                                    int $page = 1,
-                                    string $sort = '-created_at',
-                                    bool $calc_total = false): UserReportTypesResponse
+    public function userReportTypes(?UserReportTypesParams $params = null): UserReportTypesResponse
     {
+        $query = $this->getListQueryProperties(['_can_generate' => 'false']);
+
+        // Modify query, if needed
+        if ($params instanceof UserReportTypesParams) {
+            if (is_bool($with_content = $params->isWithContent())) {
+                $query['_content'] = $with_content ? 'true' : 'false';
+            }
+
+            if (is_string($filter_query = $params->getQuery())) {
+                $query['_query'] = $filter_query;
+            }
+
+            if (is_int($per_page = $params->getPerPage())) {
+                $query['_size'] = \max(1, $per_page);
+            }
+
+            if (is_int($offset = $params->getOffset())) {
+                $query['_offset'] = $offset;
+            }
+
+            if (is_int($page = $params->getPage())) {
+                $query['_page'] = \max(1, $page);
+            }
+
+            if (is_string($sort_by = $params->getSortBy())) {
+                $query['_sort'] = $sort_by;
+            }
+
+            if (is_bool($is_calc_total = $params->isCalcTotal())) {
+                $query['_calc_total'] = $is_calc_total ? 'true' : 'false';
+            }
+
+            // Unique property for `user/reports`
+            if (is_bool($can_generate = $params->isCanGenerate())) {
+                $query['_can_generate'] = $can_generate ? 'true' : 'false';
+            }
+        }
```

Comment:
If we set some common interface (like a `ItemsListInterface`) to `UserReportsParams` and `UserReportTypesParams`, we can move two large chunks of code into a private/protected method.

Methods `userReports` and `userReportTypes` will become shorter and *more-DRY*.


@tarampampam What do you think?